### ISSUE TITLE
"Stateless" public name

### ISSUE THREAD

**Issue by ghedo** [martinthomson/ech-pnmasq#2]
Section 1.1 says:
> In the ideal form of this approach, a unique public name is used for each client.

I think this would be impactical for large deployments (particularly considering these names and corresponding keys would need to be shared and synced between DNS and TLS stacks).

Now, this might be going too far, but in theory I imagine a DNS responder could generate a "single-use" (leaving DNS caching aside) RPK based on a random nonce and a secret pre-shared with whatever TLS terminating service. The random nonce could then be returned as part of the DNS response along with `spki_hash` of the RPK, and then reflected by the client to the TLS server, which could then generate the same key based on the same nonce and pre-shared secret.

This could require less state to be shared between DNS and TLS and allow for much more frequent public name rotation.

Though it could potentially be a separate extension to be used along with `public_name_authn`.

**Comment by martinthomson** [comment]
You hit a work in progress.  That's more or less exactly the design that I have in mind here.  Except that the public name is the nonce.  I have most of that design written down now, but it turns out to be fairly tricky to get right.

The challenge is in ensuring that a public name doesn't point straight at a single hidden name.  A "single-use" name can have that effect if your adversary is able to get the name, which is something that DNS caches tend to ensure will happen.  A true single use name wouldn't have this problem, but that would require the cooperation of recursive resolvers[^1].

[^1]: There's an potential opportunity for a DNS extension here, maybe.  I predict that it would encounter some resistance from the community as it depends on disabling all of the benefits of caching.  Caching is pretty important to DNS, so disabling it would be a big decision.

**Comment by dennisjackson** [comment]
Is it practical to have the client pick the public name rather than the server? This allows different clients to use different cover name selection strategies and means all the server has to do is generate an SPKI and a nonce, no complicated policy decisions for server providers on what cover names to pick from :-). 

You could either use the PSK Token mode that ghedro suggested, a lighter nonce-based mode or use a unique per-connection token e.g. `Hash(hash_spki_From_DNS_HTTPS_RR || TLS_Client_Random)` carried in the ECH Extension. 

Throwing in the Client Random means that the caching problem goes away entirely and I guess the cost of a few SHA-256 invocations is pretty marginal compared to all the HPKE and HKDF calls flying around. It would require a small number of RPKs to be in use on the server-side though which might be problematic.

**Comment by martinthomson** [comment]
I don't think that it is feasible to allow clients to pick the name.  If you consider that there is a limited surface in the outer ClientHello that the server can use to select the right first-order keys, then use those keys to pick the profile that is in use, the options are limited.

You could add another layer of public key encryption if your goal was to build an infinite stack of turtles.  So maybe that isn't an realistic option.

The challenge is that the client needs to be given information that effectively selects a server-side profile.  There are two cases to consider:

1. Where the client receives information that the adversary doesn't have.  This is an ideal state, but one that shared caching makes virtually impossible if you are getting configurations from DNS.
2. Where the client receives information that the adversary also has.  Any symmetric operations that a client might use will be as easy for the adversary to undo as it is for the server.  Your idea about hashing the SPKI is workable for the server only if the server has a limited set of SPKI values to check.  But an adversary will likely be able to recover that same set of values as well and they have a distinct advantage in terms of recovering them.

It occurs to me that the first is unlikely with SVCB, but 100% guaranteed with a retry configuration.  That makes this design highly desirable if you can get ECH configurations using a retry.  That is, if you can ever get into that state without revealing the hidden name to an adversary in the process.

That first connection is really the hard part of the problem.  It does mean that return visits to a site can be very well protected.  High value sites might even want to force clients to retry on first contact just to avoid that (which is directly opposite to what Bas suggests in #3).  It's a form of leap of faith, where a compromise on that first connection has consequences, but attackers cannot subsequently attack connections thereafter.
